lib/Sema/TypeCheckDecl.cpp

```diff
@@ -953,6 +953,61 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
           continue;
       }
 
+      // If the conflicting declarations have non-overlapping availability and
+      // - one throws and the other does not,
+      // - or they are initializers with different failability
+      // let them go.
+      bool isAcceptableVersionBasedChange = false;
+      {
+        auto *currentInit = dyn_cast<ConstructorDecl>(current);
+        auto *otherInit = dyn_cast<ConstructorDecl>(other);
+        if (currentInit && otherInit &&
+            ((currentInit->getFailability() == OTK_None) ^
+             (otherInit->getFailability() == OTK_None))) {
+          isAcceptableVersionBasedChange = true;
+        }
+      }
+      {
+        auto *currentAFD = dyn_cast<AbstractFunctionDecl>(current);
+        auto *otherAFD = dyn_cast<AbstractFunctionDecl>(other);
+        if (currentAFD && otherAFD &&
+            currentAFD->hasThrows() != otherAFD->hasThrows()) {
+          isAcceptableVersionBasedChange = true;
+        }
+      }
+      if (isAcceptableVersionBasedChange) {
+        struct AvailabilityRange {
+          Optional<clang::VersionTuple> introduced;
+          Optional<clang::VersionTuple> obsoleted;
+        };
+
+        auto findSwiftAvailability =
+            [](const ValueDecl *VD) -> AvailabilityRange {
+          AvailabilityRange result;
+          for (auto *attr : VD->getAttrs().getAttributes<AvailableAttr>()) {
+            if (attr->PlatformAgnostic ==
+                  PlatformAgnosticAvailabilityKind::SwiftVersionSpecific) {
+              if (attr->Introduced)
+                result.introduced = attr->Introduced;
+              if (attr->Obsoleted)
+                result.obsoleted = attr->Obsoleted;
+            }
+          }
+          return result;
+        };
+
+        AvailabilityRange currentAvail = findSwiftAvailability(current);
+        AvailabilityRange otherAvail = findSwiftAvailability(other);
+        if (currentAvail.introduced && otherAvail.obsoleted &&
+            *currentAvail.introduced >= *otherAvail.obsoleted) {
+          continue;
+        }
+        if (otherAvail.introduced && currentAvail.obsoleted &&
+            *otherAvail.introduced >= *currentAvail.obsoleted) {
+          continue;
+        }
```

Comment:
minor suggestion: consider moving this logic into a helper, to the effect of `bool AvailabilityRange::precedes(const AvailabilityRange&) const` (there probably exist better naming choices, maybe even `operator<`)